subgraph/packages/subgraph-dufour-channel/src/channels.ts

```diff
@@ -0,0 +1,145 @@
+    }
+
+    channel.balance = convertEthToDecimal(event.params.newBalance)
+    channel.lastOpenedAt = event.block.timestamp
+    channel.save()
+}
+
+export function handleChannelClosed(event: ChannelClosed): void {
+    let channelId = event.params.channelId.toHex()
+    let channel = Channel.load(channelId)
+
+    if (channel == null) {
+        log.error("Closing non-existing channel", [])
+        return
+    }
+
+    channel.lastClosedAt = event.block.timestamp
+    channel.status = convertStatusToEnum(2)
+    channel.balance = zeroBD()
+    channel.save()
+}
+
+export function handleChannelOpened(event: ChannelOpened): void {
+    // source and destination channel
+    let sourceId = event.params.source.toHex()
+    let destinationId = event.params.destination.toHex()
+
+    let source = getOrInitiateAccount(sourceId)
+    let destination = getOrInitiateAccount(destinationId)
+
+    // channelId
+    let channelId = getChannelId(event.params.source, event.params.destination).toHex()
+
+    source.fromChannelsCount = source.fromChannelsCount.plus(oneBigInt())
+    source.save()
+
+    destination.toChannelsCount = destination.toChannelsCount.plus(oneBigInt())
+    destination.save()
+
+    let channel = initiateChannel(channelId, sourceId, destinationId)
+    channel.lastOpenedAt = event.block.timestamp
+    channel.save()
+}
+
+
+export function handleOutgoingChannelClosureInitiated(event: OutgoingChannelClosureInitiated): void {
+    let channelId = event.params.channelId.toHex()
+    let channel = Channel.load(channelId)
+
+    if (channel == null) {
+        log.error("Initiate outgoing channel closure for non-existing channel", [])
+        return
+    }
+
+    channel.status = convertStatusToEnum(2)
+}
+
+export function handleTicketRedeemed(event: TicketRedeemed): void {
+    // get the channel epoch, which is not part of the event
+    let channelId = event.params.channelId
+
+    // let ticketEpoch = event.params.ticketEpoch
+    let ticketIndex = event.params.newTicketIndex
+
+    // create new ticket
+    let ticketId = channelId.toHex() + "-" + ticketIndex.toString()
+    // let ticketId = channelId.toHex() + "-" + channelEpoch.toString() + "-" + ticketEpoch.toString() + "-" + ticketIndex.toString()
+
+    let ticket = Ticket.load(ticketId)
+
+    if (ticket == null) {
+        log.error("Redeem non-existing ticket", [])
+        return
+    }
+
+    ticket.redeemedAt = event.block.timestamp
+
+    // TODO: Missing proofOfRelaySecre
+    ticket.save()
+
+
+    // update channel
+    let channel = Channel.load(channelId.toHex())
+    if (channel == null) {
+        log.error("Redeem a ticket for non-existing channel", [])
+        return
+    }
+
+    channel.redeemedTicketCount = channel.redeemedTicketCount.plus(oneBigInt())
+    channel.ticketEpoch = ticket.ticketEpoch
+    channel.ticketIndex = ticket.ticketIndex
+
+    channel.save()
+}
+
+export function handleRedeemTicketCall(call: RedeemTicketCall): void {
```

Comment:
will subgraph also index failed calls? or should the failed calls handled by the subgraph creator?

Reply:
As it's the only place where I manage to find many of the parameters, I guess we need to listen to the redeemticket call. However, this call doesn't track if the call fails, or I cannot find it.